### ISSUE TITLE
[FEATURE REQUEST] cv2.MORPH_ELLIPSE should give a rotationally symmetric kernel for square kernel sizes

### ISSUE THREAD

**Issue by alexander-soare** [opencv/opencv#22042]
`cv2.MORPH_ELLIPSE` is not rotationally symmetrical when the kernel is square. For me, and I'm assuming for others too, this is counter-intuitive.

```
>>> cv2.getStructuringElement(cv2.MORPH_ELLIPSE, [5, 5])
array([[0, 0, 1, 0, 0],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1],
       [0, 0, 1, 0, 0]], dtype=uint8)
```

I would expect the behaviour to look more like that of scikit-image. 

```
>>> skimage.morphology.disk(2)
array([[0, 0, 1, 0, 0],
       [0, 1, 1, 1, 0],
       [1, 1, 1, 1, 1],
       [0, 1, 1, 1, 0],
       [0, 0, 1, 0, 0]], dtype=uint8)
```

**Comment by catree** [comment]
Matlab also outputs a symmetrical kernel:

```
SE3 = strel('disk', 2)
disp(SE3.Neighborhood)
```

You can run this code in your browser (click on `Try This Example`):

https://www.mathworks.com/help/images/ref/strel.html#bu7pnvx-1

**Comment by crackwitz** [comment]
I find using the `circle()` call for this questionable.

Just assign the kernel elements according to distance. That is symmetric and most accurate.
```
if x**2 + y**2 <= radius**2:
    kernel[i, j] = 1
```

And use half-integer radii because the results actually look round on the sides, not spiky.

radius 2.5 vs 3.0 vs  3.5 vs 4.0:

![image](https://github.com/opencv/opencv/assets/7065108/145d3bf6-6575-42df-8d1c-adc3200b4317) ![image](https://github.com/opencv/opencv/assets/7065108/4ee285ec-7e6e-4c21-b7e1-293d595f8a1b)

![image](https://github.com/opencv/opencv/assets/7065108/2c7915e6-b455-4058-ab6e-84661a5c514a) ![image](https://github.com/opencv/opencv/assets/7065108/91b2ee19-4788-4217-93d7-4fd73492ffa2)

**Comment by crisluengo** [comment]
@crackwitz If you use `x**2 + y**2 < radius**2` then the spikes go away (unless you pick weird radii just above integers, in which case you’re asking for it…).

**Comment by crackwitz** [comment]
> `<`
Yeah but the shapes for integer radii still look squarish, even without the spikes

**Comment by crisluengo** [comment]
> putting the center not in the middle of a pixel, but on the edge between four pixels
@crackwitz

That is fine for openings and closings, but not for dilations and erosions, where you want symmetry around a central pixel. For (isotropic) dilations and erosions you always need to put the center on a pixel.

You can get rid of the squarish-ness by using an hexagonal grid instead of a square grid. 😁

**Comment by crackwitz** [comment]
_Even_-sized kernels can't ever be symmetric in a way that a pixel center can be the kernel center. They'll always add a half-pixel shift to the result. If anyone requests an even-sized kernel, I think this is understood, and beyond the topic of *potato circles* ;)

Here are a few more shapes, for comparison. These use the `<=` calculation. The `<` calculation only differs in dropping pixels that are precisely on the edge, i.e. those spikes in cardinal directions.

For odd diameters (half-integer radii), an integer anchor works fine.

The spikes for exact radius distance only occur in even diameters (integer radii) and integer anchor. Even without those spikes, the remaining blob looks more square than the alternative that uses a half-integer anchor.

![image](https://github.com/opencv/opencv/assets/7065108/e241edeb-d62d-42c9-867b-7e429e40c509)

![image](https://github.com/opencv/opencv/assets/7065108/f89ca1c3-7c57-4a5c-84f1-e6ae2ebcb7ce)

**Comment by alexander-soare** [comment]
@crackwitz I think that really works. Using diameter is better than using radius as a parameter because then I don't have to think too hard about what is meant by radius. In your proposed scheme, the diameter is always the number of pixels counted along the horizontal or vertical axis of symmetry which is totally intuitive. With erosion and dilation I usually think of a horizontal or vertical edge and how many pixels I want to see that change by. With your formulation I would always pick odd sized kernels and I'd know its (d - 1)/2. Easy :)

**Comment by crackwitz** [comment]
erosion/dilation are ok with even-sized kernels too. the default anchor gets rounded to any of the pixel positions close to the true center. those operations, on even kernels, would introduce a shift of half a pixel, strictly speaking.

`morphologyEx` appears to be dumb. it produces two half-integer shifts in the same direction.
```python
>>> a = np.zeros((10,10), 'u1')
>>> a[2:4,2:4] = 99
>>> a
array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 99, 99,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 99, 99,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0]], dtype=uint8)
>>> a = np.zeros((10,10), 'u1')
>>> a[4:6,4:6] = 99
>>> a
array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 99, 99,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 99, 99,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0]], dtype=uint8)
>>> kernel = np.ones((2,2), 'u1') * 255
>>> kernel
array([[255, 255],
       [255, 255]], dtype=uint8)
>>> cv.morphologyEx(src=a, op=cv.MORPH_CLOSE, kernel=kernel)
array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 99, 99,  0,  0,  0],
       [ 0,  0,  0,  0,  0, 99, 99,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0]], dtype=uint8)
```

**Comment by alexander-soare** [comment]
Hm, well one could think of closing as "erode then dilate" but (and maybe I'm wrong) I like to think of it as "everywhere you can fit the kernel gets to stay". The latter wouldn't cause a shift, even if there is no center pixel.

**Comment by crackwitz** [comment]
erode, then dilate is opening, actually, but anyway, I've been editing my last post a few times, so I hope it's now presenting everything better than the first version.

**Comment by alexander-soare** [comment]
I guess then, since my mental shortcut ("everywhere you fit the kernel gets to stay") is just not true, then the shift is not "unexpected". As an aside, I think my definition describes a more useful operation.

**Comment by crackwitz** [comment]
> "everywhere you can fit the kernel gets to stay"

that is true for erosion. erosion by a 2x2 kernel would reduce a 2x2 connected component to 1x1. dilation is the opposite ("any") and produces the 3x3 component.

**Comment by alexander-soare** [comment]
Nope, I really was referring to closing. Like a 3x3 patch would stay with a 3x3 kernel because the kernel fits inside it, whereas anything smaller would disappear. But yeah like I said, that was an aside.

**Comment by crackwitz** [comment]
> anything smaller would disappear

that would be opening, not closing, traditionally, "white" pixels are true/foreground and black pixels are false/background. I think I only confused you by trying to correct you on that.

**Comment by crisluengo** [comment]
Nice! It looks like OoenCV’s closing is broken (and we can assume that the opening is too).

let me clarify a few things, you both are more or less right. The closing removes all the holes whereeverything

**Comment by crisluengo** [comment]
[If you’re reading through email, please click on the link, I posted the previous comment before finishing writing it — fat fingers on a small phone screen.]

@crackwitz ‘s kernels look very nice. I said that it’s not OK for dilations and erosions because of the shift you get, which I wouldn’t expect. But maybe people would expect the shift if you think of the diameter as the side length of the square kernel that the circle is drawn in.

Those kernels would be great — but only if the opening and closing are fixed to properly mirror the kernel before applying the second operator.

**Comment by alexander-soare** [comment]
@crisluengo thanks for bringing us on the same page! And I was going to say something about mirroring but I thought I'd be reaching too far. Glad to know it's actually right. To mop up my mess:

closing = dilate then erode = only keep background that the kernel can fit in
opening = erode then dilate = only keep foreground that the kernel can fit in
